tck/src/main/java/org/eclipse/microprofile/lra/tck/participant/api/AfterLRAListener.java

```diff
@@ -0,0 +1,86 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.eclipse.microprofile.lra.tck.participant.api;
+
+import org.eclipse.microprofile.lra.annotation.LRAStatus;
+import org.eclipse.microprofile.lra.annotation.ws.rs.AfterLRA;
+import org.eclipse.microprofile.lra.annotation.ws.rs.LRA;
+import org.eclipse.microprofile.lra.tck.service.LRAMetricService;
+import org.eclipse.microprofile.lra.tck.service.LRAMetricType;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import javax.ws.rs.HeaderParam;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_CONTEXT_HEADER;
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_ENDED_CONTEXT_HEADER;
+import static org.eclipse.microprofile.lra.annotation.ws.rs.LRA.LRA_HTTP_RECOVERY_HEADER;
+
+/**
+ * resource for testing that methods annotated with
+ * {@link AfterLRA}
+ * are notified correctly when an LRA terminates
+ */
+@ApplicationScoped
+@Path(AfterLRAListener.AFTER_LRA_LISTENER_PATH)
+public class AfterLRAListener {
+    public static final String AFTER_LRA_LISTENER_PATH = "after-lra-listener";
+    public static final String WORK = "work";
+
+    private static final String AFTER_LRA = "/after";
+
+    @Inject
+    private LRAMetricService lraMetricService;
+
+    @PUT
+    @Path(WORK)
+    @LRA(value = LRA.Type.REQUIRED, end = false)
+    public Response activityWithLRA(@HeaderParam(LRA_HTTP_RECOVERY_HEADER) URI recoveryId,
+                                    @HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId) {
+        return Response.ok().build();
+    }
+
+    @PUT
+    @Path(AFTER_LRA)
+    @AfterLRA // this method will be called when the LRA associated with the method activityWithLRA finishes
+    public Response afterEnd(@HeaderParam(LRA_HTTP_ENDED_CONTEXT_HEADER) URI lraId, String endStatus) {
+        LRAStatus status = LRAStatus.valueOf(endStatus);
+
+        if (status.equals(LRAStatus.Closed)) {
+            lraMetricService.incrementMetric(LRAMetricType.Closed, lraId,
+                    AfterLRAListener.class.getName());
+        } else if (status.equals(LRAStatus.Cancelled)) {
+            lraMetricService.incrementMetric(LRAMetricType.Cancelled, lraId,
+                    AfterLRAListener.class.getName());
+        } else if (status.equals(LRAStatus.FailedToCancel)) {
+            lraMetricService.incrementMetric(LRAMetricType.FailedToCancel,
+                    lraId, AfterLRAListener.class.getName());
+        } else if (status.equals(LRAStatus.FailedToClose)) {
+            lraMetricService.incrementMetric(LRAMetricType.FailedToClose,
+                    lraId, AfterLRAListener.class.getName());
+        }
+
+        return Response.ok().build();
```

Comment:
Describe this in the spec and what happens when endpoint is called within an invalid body (not a LRAStatus)

Reply:
good catch, should be in the spec text - 200 does not seem the right return code

Reply:
It would be a spec violation if the implementation calls the endLRA notification with an invalid status.

The spec is not interested in the result (it is similar to @Forget in this respect). In fact a 200 response would tell the caller that the listener successfully received the notification.

Reply:
400?

Reply:
We need to mention in the spec text then, that `@AfterLRA` is on the best-effort basis. This, however, seems that some business may depend on this functionality so maybe we should think more about whether it should be best-effort? Please don't take this as something to stall this PR, we can revisit it in 1.x, I just wanted to point it out.

Reply:
Then I will include the following text in the spec for this feature:

 "the notification is guaranteed to eventually be delivered and if the callback returns anything other that 200 then the implementation will keep retrying."